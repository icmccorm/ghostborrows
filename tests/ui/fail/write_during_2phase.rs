use ghostborrows::*;

fn main() {
    /*
     
     // Not UB according to Stacked Borrows.
     // Should be UB according to Tree Borrows.

     fn write_during_2phase() {
        let x = &mut 0u8;
        let xraw = x as *mut u8;
        print(
            x,
            unsafe { *xraw += 1 };
        )
     }
    
    */

    let value = Value::new(0);
    value.borrow_mut(|x, token| {
        let (pointer, perm) = x.split();
        x.borrow_mut(token, |_, _| {
            *pointer.as_mut(perm) += 1;
        });
    });
}