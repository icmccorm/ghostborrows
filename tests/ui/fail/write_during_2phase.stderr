error[E0308]: mismatched types
  --> tests/ui/fail/write_during_2phase.rs:24:29
   |
24 |             *pointer.as_mut(perm) += 1;
   |                      ------ ^^^^ expected `&mut dyn AllowsWrite<'_, _>`, found `Reserved<'_, {integer}>`
   |                      |
   |                      arguments to this method are incorrect
   |
   = note: expected mutable reference `&mut dyn AllowsWrite<'_, _>`
                         found struct `Reserved<'_, {integer}>`
note: method defined here
  --> src/refs.rs
   |
   |     pub fn as_mut<T>(&self, _: &mut dyn AllowsWrite<'tag, T>) -> &mut T {
   |            ^^^^^^
